### ISSUE TITLE
centernet as rpn

### ISSUE THREAD

**Issue by alexwq100** [xingyizhou/CenterNet#453]
hi,@xingyizhou
   have you tried to use centernet as rpn which means refactoring it to two stage detector or even add mask branch? I tried to do that ,but the performance doesn't do better than the maskrcnn which use resnet50 as backbone and  slightly better than centernet  origin version. 
Hope to your comments. thks

**Comment by xingyizhou** [comment]
Hi alexwq100,
  Thanks for sharing this result! I am interested in doing that but haven't got time. One problem is which backbone to use for the two-stage model. Are you using dla34 as backbone or resnet50? I have tried res50 (with naive upsample or IDA upsample) for centernet and dla34 for RetinaNet, but both end up in a dropped performance than the default one ...

**Comment by alexwq100** [comment]
yes ,I use dla34 as the backbone , I tried two methods, one is map all the proposals to the feature map which you use to train the centernet in the paper ,another way is use x in the the https://github.com/xingyizhou/CenterNet/blob/master/src/lib/models/networks/pose_dla_dcn.py#L472  , which has stride 4,8,16 and 32 and it is consistent with the stride of fpn output of resnet 50, but the result not get much better and it is needed quite tuning.

**Comment by fafancier** [comment]
I also add  roi-align operation using the predicted boxes , then using RCNN header to repredict the box offset. I just do a rough experiment. I use ResNet101 as the backbone,  pool roies from the conv4 (H/16,W/16 layer), which is later proved to be not sufficient for predict, and for the reason that RCNN also use resnet  layer4.  so  sometimes , it  got the same wrong  prediction. But, I found  that the training procedure help the centernet's single scale prediction  on ResNe101  impove a little mAP on coco validation, the loss of hm, wh got a little decreased.

**Comment by alexwq100** [comment]
hi,@fafancier, you said "But, I found that the training procedure help the centernet's single scale prediction on ResNet101 " ,I am not very clear about that, what the single scale do you mean?

**Comment by fafancier** [comment]
hi,@alexwq100 
Actually single scale test is not using multi-sacle image test which I didn't test. I just add the roi align and using conv5 head for classification then train 25epoch and the mAP increase a little. the green line circled change to 35.2 and 54.8.  maybe  it  just  converge  a little  more.




![未命名图片](https://user-images.githubusercontent.com/39083471/68537856-d5392280-03a5-11ea-8676-ea6f8f162c8c.png)


add rcnn head then  train 25 epoch result.
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.352
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.548
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.373
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.151
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.392
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.523
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.301
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.469
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.485
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.238
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.540
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.708

**Comment by xingyizhou** [comment]
Thanks @fafancier . These are very interesting results. I would say applying ROI align after the upconv layers in DLA-up or upsampled ResNets makes more sense, as the largest feature map in FPN is also in stride 4.

**Comment by TWDH** [comment]
@fafancier Hi, I'm interested in your idea, did you add the RCNN and ROI align in parallel with the heatmap,wh and offset? like how do you use the RCNN to improve the performance of heatmap loss, I don't really get it. 
Furthermore, can you tell me how do you train another 20 epoch based on the original model?
Looking forward to your reply!

**Comment by TWDH** [comment]
@xingyizhou @fafancier I am very interested in your result and implement my own one using RCNN header to repredict bbox offset and classification (also added roi align)but result becomes lower than the default one by using the dla backbone.  Do you add Rcnn-bbox-offset and Rcnn-bbox-classifications loss directly to the whole loos function? or just use the classification loss to help with the original CeterNet prediction? Thanks very much!

**Comment by fafancier** [comment]
Hi, @TWDH I've tried several combinations and do much experiment on backbone ResNet101. It turns out that It seems better to use wh without bbox regression. and the classification actually can not catch up with the score of original CenterNet.  Maybe it's the problem of my trainning and parameter tuning skill. But, training with ROI Aligned map and Cls loss the original CenterNet will get a little boost.    I also find that CenterNet produces much low score output, which makes the results' json file is large. however, the RCNN header is smaller. Besides, removing low scores in CenterNet will drop the accuracy in evaluation.  CenterNet is easy to produce two identity center false positive between the confusing object.

**Comment by TWDH** [comment]
> Hi, @TWDH I've tried several combinations and do much experiment on backbone ResNet101. It turns out that It seems better to use wh without bbox regression. and the classification actually can not catch up with the score of original CenterNet. Maybe it's the problem of my trainning and parameter tuning skill. But, training with ROI Aligned map and Cls loss the original CenterNet will get a little boost. I also find that CenterNet produces much low score output, which makes the results' json file is large. however, the RCNN header is smaller. Besides, removing low scores in CenterNet will drop the accuracy in evaluation. CenterNet is easy to produce two identity center false positive between the confusing object.

Hi, @fafancier, Thanks so much for your reply! I am not so sure what do you mean by "training with ROI Aligned map and cls loss", Do you mean we use to original CenterNet output (box) as ROIs and let it go through the ROI Aligned map and only use cls loss when training? (like, loss = hm_loss + wh_loss + offset_loss + RCNN_cls_loss) and it will boost the original CenterNet output, right? Not sure if I understand it correct, cause I use rcnn_bbox and rcnn_cls to repredict the box and results in a decrease of 1.3% compared to the original CenterNet. 
Plus, do you mind we become friends on Facebook(He Zhu) or Snapchat(hezhukevin) so that we can discuss it? Looking forward to your reply.

**Comment by fafancier** [comment]
@TWDH  yes, I try loss = hm_loss + wh_loss  + RCNN_cls_loss. with decrease also. I am not sure whether the offset loss will be  helpful.

**Comment by TWDH** [comment]
@fafancier For the condition that you said accuracy got increased, is the loss = hm_loss + wh_loss + RCNN_cls_loss? or other loss combinations... Thanks :) Cause I got confused you mentioned using this loss combination increases in the first comment and then "with decrease also" for the last comment...

**Comment by fafancier** [comment]
@TWDH  actually, the loss help refine the feature which helps the original network increase. The RCNN head's classification experiment is not better than origin.

**Comment by TWDH** [comment]
@fafancier So, is that you mean the no matter if you add RCNN_cls_loss to the whole loss or not doesn't really matter, on earth the loss will decrease to refine the prediction? the RCNN header is totally useless?

**Comment by fafancier** [comment]
@TWDH adding RCNN_cls_loss to the whole loss helps improve CenterNet in my experiment

**Comment by TWDH** [comment]
@fafancier Thanks bro, I was wondering how do you get prediction results from original centernet? I just use the functions in "base_detection.py" to predict every training data images like in "demo.py" and only choose the bbox with confidence greater than 0.3 (which should have 100 bbox predictions in default), I'm not sure if this will affect the evaluation result? cause theoretically, bbox regression is able to refine the RPN results anyway ... it just make me confused. Looking forward to your comments!

**Comment by TWDH** [comment]
Hi @fafancier,sorry to borther you again. I tried your method with loss = hm_loss + wh_loss + offset_loss + RCNN_cls_loss with dlaNet as backbone and 2 FC layer(1024 channels) as RCNN header. But my experiment doesn't show good result, **the training loss consistently goes down while validation loss(hm, wh and offset) goes up which does harm to the original CenterNet** instead of refining it, do you experience the same situation? I use learning rate at 1.24e-4 and decrease by a factor of 0.1 at epoch 10. Would you please tell me what lr did you use? may be comment on it, Thanks a lot 
[RESULT.txt](https://github.com/xingyizhou/CenterNet/files/4718035/RESULT.txt)
